rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,716 @@
+   rejoining conditions in tandem with the rejoin counter. The
+   `spec.rejoining.must_rejoin_before` field may accomplish this.
+
+TODO: Ensure this is solved and not confusing.
+
+#### Keypair Rotation
+
+Given the long-lived nature of the keypair credential, it's important to support
+rotation without bot downtime. Ideally, it should be possible to initiate a
+rotation from either the server (e.g. by setting a `rotate_on_next_renewal` flag
+on the token/bot instance) or `tbot` client.
+
+TODO: Expand this.
+
+#### Remaining Downsides
+
+- Repairing a bot that has exhausted all of its rejoins is still a semi-manual
+  process. It is significantly easier, and does not necessarily require any
+  changes on the impacted bot node itself, but is still annoying. Users can opt
+  out of this by setting `.spec.rejoining.unlimited=true`, but this has obvious
+  security implications.
+
+- Effort required to configure IaC / Terraform is still fairly high, even if
+  reduced.
+
+### Other Supporting Improvements
+
+Alongside `bound-keypair` joining, we have several UX proposals to further
+improve the usability of non-delegated joining.
+
+#### Longer-Lived Bots
+
+The renewable identity's 24 hour maximum TTL is too restrictive and should be
+lengthened. We propose raising this limit to 7 days, but keeping the default
+(1hr) the same.
+
+#### Bot Instance Locking
+
+When we introduced bot instances in
+[RFD0162](0162-machine-id-token-join-method-bot-instance.md), it allowed many
+bot instances to join under a single bot (and Teleport user). Generation counter
+checks were moved out of a bot's user and into bot instances, however when a
+generation counter mismatch occurs, the resulting lock is still filed against
+the user as a whole. This means a generation counter lockout in one instance can
+easily impact all other instances under the same bot.
+
+We can introduce several new lock targets to address this:
+
+- Bot instance UUID locking: prevent access by a particular instance. This will
+  not lock bots that have since reauthenticated and received a new bot instance
+  UUID.
+
+- Join token locking: locks all bot instances that joined with a particular
+  token. This may require introduction of a new certificate field to track the
+  exact join token used.
+
+- Public key locking: locks bots joining with a particular public key. A
+  compromised bot could theoretically generate a fresh keypair so a join token
+  lock is the primary locking solution, however this will prevent joining with
+  another token.
+
+#### Bot Joining URIs
+
+Instead of separate flags for proxy, join method, token value, and other
+potential join-method specific parameters, we propose adopting joining URIs.
+These would provide a single value users can copy to pre-fill various
+configuration fields and would greatly improve the onboarding experience.
+
+The URI syntax might look like this:
+```
+tbot+[auth|proxy]://[join method]:[token value]@[addr]:[port]?key=val&foo=bar
```

Comment:
For easy of copying this, you could base64 encode this and/or even sign it like a JWT.

Reply:
I think there are some benefits to it being cleartext, mainly ease of confirming bots are routed to the right place, and ability to tweak the URI if needed, e.g. connecting to a leaf cluster. From a UX perspective, it's still functionally a single token for copy and paste purposes, which Teleport can fully compose it for you in the CLI/web. Do you see a use case for shorter connection strings? 

I'm not too sure of the value of signing these - could you elaborate on the use case a bit? Bots still verify TLS on startup, and in general I don't see bots connecting to a hostile Teleport instance as a meaningful threat vector.

Reply:
A token is easily copy-pastable in a shell (i.e. vim/bash etc), where a complex url usually is a bit more annoying. Especially if you have `+` and `&` characters.
Having those replaced via base64 encoding significantly reduces the amount of user error in most cases.

Signing would just be to validate that you have copied the whole string. Maybe I should have called it hashing.